8277309: Add support for H.265/HEVC to HTTP Live Streaming

sashamatveev opened the pull request:
- Added support for fragmented MP4 with HEVC/H.265, H.264 and AAC. 
 - Added support for elementary AAC streams without any container for audio only streams.
 - Added "aacparse" plugin from GStreamer. Required on Linux, since decoder cannot handle AAC elementary streams directly. DirectShow decoder works without it.
 - DirectShow H.264 decoder on Windows and H.265/H.264 decoder on Linux will be reloaded when fMP4 stream changes resolution. Dynamic format change did not worked for these streams on Windows and Linux.

kevinrushforth commented:
/reviewers 2

yososs commented:
Why not use BitSet instead of ArrayList<Boolean>?

yososs commented:
modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/HLSConnectionHolder.java

I noticed that in this change, ArrayList\<Boolean> is rewritten as ArrayList\<>.
The BitSet rewrite needs some thought, so it might be a good idea to make it a separate issue.
In the file I pointed out, it is used in many more places than the one I pointed out.

kevinrushforth commented on modules/javafx.media/src/main/native/gstreamer/plugins/dshowwrapper/dshowwrapper.cpp:
```diff
@@ -1931,9 +1939,66 @@ static void dshowwrapper_load_decoder_h264(GstDShowWrapper *decoder, JFX_CODEC_I
+        decoder->pMediaControl->GetState(5000, &fs);
+    }
+
+    if (decoder->pDecoder != NULL && decoder->pGraph != NULL)
```
Can `decoder->pDecoder` be non-null and `decoder->pGraph` be null? If so, then the decoder wouldn't be released. Does it need to be?

sashamatveev replied:
Unlikely. I changed it to check for `if (decoder->pGraph == NULL)` at beginning of function and return `FALSE` if it happens, since we cannot swap decoder if graph is not initialized. Not sure if it can happen, so this check is more for just in case. Yes, we need to release decoder and re-create it during format change. I tried all possible ways to change format dynamically and nothing worked. Only complete recreation of decoder instance worked.

kevinrushforth reviewed (approved):
All my testing looks good. I spot checked the new code, and it looks good as well. I left a couple questions for you, but I'll approved it as is.

sashamatveev commented on modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/HLSConnectionHolder.java:
```diff
@@ -274,58 +349,64 @@ private void stateInit() {
+                if (isFragmentedMP4()) {
+                    mediaFileIndex = 0;
+                }
+            } catch (Exception e) {
```
No. Removed.

sashamatveev commented:
Fixed several issues found during additional testing:
- sendHeader was true by default which caused sending first and second segment together for MP2TS and MP3/AAC elementary streams. This variable should be true only for fragmented MP4.
- Removed G_SIGNAL_NO_RECURSE flag from javasource signals. From GLib doc "G_SIGNAL_NO_RECURSE - Signals being emitted for an object while currently being in emission for this very object will not be emitted recursively, but instead cause the first emission to be restarted." Basically it caused skipping data segments since in some cases loadNextSegment() was called twice. We do use signal mechanism from separate thread when reading duration. This issue is intermittent. All GStreamer provided elements do not use G_SIGNAL_NO_RECURSE flag, so not sure why javasource was using it.
- Fixed duration calculation. We should not count duration of header, since header does not contain any data.
- Added code which will adjust PTS during rendering to start with 0, otherwise (after duration fix) we will drop frames with PTS > duration. Note: In fMP4 HLS streams PTS starts with N and goes upto duration + N. Our rendering code expects start with 0 and stop at duration.
- Minor code cleanup and removal of unnecessary "if (segment.start != 0)" check in hlsprogressbuffer.

johanvos commented:
This looks really great. I'm doing a few more tests before approving, but I don't expect issues.
I wonder if we can add some simple resources that allow testing the different protocols on different platforms?

kevinrushforth commented:
> I wonder if we can add some simple resources that allow testing the different protocols on different platforms?

I can't think of a good way to do that, especially for this feature, since it requires a web server.

kevinrushforth commented:
Interesting idea. Btw, we don't have any media tests at all in our repo at present, so there is a more basic issue that would need to be solved. We could file an RFE, but there are some challenges in doing anything that is useful enough to justify the effort.
